### PR TITLE
Fb1322

### PR DESCRIPTION
**Pull request by spiresheep**

### CODE REVIEW

File: source/array_table_model.ts [hunk 1]
```diff
@@ -55,32 +58,82 @@ export class ArrayTableModel extends TableModel {
    * @throws RangeError - The source or destination are not within this table's
    *                      range.
    */
-  public moveRow(source: number, destination: number): void {}
+  public moveRow(source: number, destination: number): void {
+    if(source > this.rowCount || source < 0) {
```
Comment (kranar) [inline source/array_table_model.ts]: If ```source === this.rowCount``` then you should also throw an exception.

File: source/array_table_model.ts [hunk 2]
```diff
@@ -55,32 +58,82 @@ export class ArrayTableModel extends TableModel {
...
+      throw RangeError();
+    }
+    if(destination > this.rowCount || destination < 0) {
```
Comment (kranar) [inline source/array_table_model.ts]: Same as above.

File: source/array_table_model.ts [hunk 3]
```diff
@@ -55,32 +58,82 @@ export class ArrayTableModel extends TableModel {
...
+      throw RangeError();
+    }
+    if(source === destination) {
+      return;
+    }
+    this.beginTransaction();
+    const row = this.values[source];
+    this.values.splice(source, 1);
+    this.values.splice(destination, 0, row.slice());
```
Comment (kranar) [inline source/array_table_model.ts]: Don't copy the ```row```.

File: source/array_table_model.ts [hunk 4]
```diff
@@ -55,32 +58,82 @@ export class ArrayTableModel extends TableModel {
...
+    this.endTransaction();
+  }
 
   /** Removes a row from the table.
    * @param index - The index of the row to remove.
    * @throws RangeError - The index is not within this table's range.
    */
-  public removeRow(index: number): void {}
+  public removeRow(index: number): void {
+    if(index > this.rowCount - 1 || index < 0) {
```
Comment (kranar) [inline source/array_table_model.ts]: Use ```index >= this.rowCount``` instead of ```index > xxx - 1```.

File: source/array_table_model.ts [hunk 5]
```diff
@@ -55,32 +58,82 @@ export class ArrayTableModel extends TableModel {
...
+      this.dispatcher.dispatch(null);
```
Comment (kranar) [inline source/array_table_model.ts]: Don't dispatch a ```null```. A subscriber doesn't care about an error.

File: source/array_table_model.ts [hunk 6]
```diff
@@ -55,32 +58,82 @@ export class ArrayTableModel extends TableModel {
...
+      throw RangeError();
+    }
+    this.beginTransaction();
+    const row = new ArrayTableModel();
+    row.values.push(this.values.slice(index, index + 1));
```
Comment (kranar) [inline source/array_table_model.ts]: Don't copy the row being removed.

File: source/array_table_model.ts [hunk 7]
```diff
@@ -55,32 +58,82 @@ export class ArrayTableModel extends TableModel {
...
-  public set(row: number, column: number, value: any): void {}
+  public set(row: number, column: number, value: any): void {
+    if(row > this.rowCount - 1 || row  < 0) {
+      throw RangeError();
+    }
+    if(column > this.columnCount - 1  || column < 0) {
+      throw RangeError();
+    }
+    this.beginTransaction();
+    const table = new ArrayTableModel();
```
Comment (kranar) [inline source/array_table_model.ts]: No update is performed here. You will want to actually modify the table itself.

File: source/array_table_model.ts [hunk 8]
```diff
@@ -55,32 +58,82 @@ export class ArrayTableModel extends TableModel {
...
+    const previous = table.values[row].slice(column, column + 1);
+    table.values[row][column] = value;
+    this.operations.push(new UpdateValueOperation(
+      row, column, previous, value));
+    this.endTransaction();
+  }
 
   public get rowCount(): number {
-    return 0;
+    if(this.values.length === 1 && this.values[0].length === 0) {
```
Comment (kranar) [inline source/array_table_model.ts]: This should strictly evaluate to ```this.values.length```

File: tests/array_table_model_tester.ts [hunk 9]
```diff
@@ -36,4 +51,165 @@ export class ArrayTableModelTester {
...
+          return;
+        } else {
+          Expect(false).toEqual(true);
+        }
+      } else {
+        Expect(false).toEqual(true);
+      }
+    };
+    const listener = model.connect(slot);
+    Expect(model.rowCount).toEqual(0);
```
Comment (kranar) [inline tests/array_table_model_tester.ts]: You don't need to duplicate all these tests. If this has been tested in another unit test then remove it from this one.

**Review by kranar — CHANGES_REQUESTED** [review]
So far so good.

File: source/array_table_model.ts (hunk 6 above)
Comment (spiresheep) [reply source/array_table_model.ts]: What should I do instead?

File: source/array_table_model.ts [hunk 10]
```diff
@@ -55,32 +58,76 @@ export class ArrayTableModel extends TableModel {
...
+  }
 
   /** Removes a row from the table.
    * @param index - The index of the row to remove.
    * @throws RangeError - The index is not within this table's range.
    */
-  public removeRow(index: number): void {}
+  public removeRow(index: number): void {
+    if(index >= this.rowCount || index < 0) {
+
```
Comment (kranar) [inline source/array_table_model.ts]: Remove blank line.

File: tests/array_table_model_tester.ts [hunk 11]
```diff
@@ -36,4 +51,149 @@ export class ArrayTableModelTester {
     Expect(receivedIndex).toEqual(1);
     listener.unlisten();
   }
+  /** Tests removing rows. */
```
Comment (kranar) [inline tests/array_table_model_tester.ts:55]: Put blank line between declarations, ie. above line 54.

**Review by kranar — CHANGES_REQUESTED** [review]

**Review by kranar — APPROVED** [review]

**Pull request merged**